File: moling-server/app/models/chapter.py
"""墨灵 (Moling) — Chapter ORM Model."""

from datetime import datetime
from typing import Optional

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text, JSON
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import BaseModel, SoftDeleteMixin


class Chapter(BaseModel, SoftDeleteMixin):
    """A single chapter within a project."""

    __tablename__ = "chapters"

    project_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
        comment="所属项目 ID",
    )
    title: Mapped[str] = mapped_column(
        String(200),
        nullable=False,
        comment="章节标题",
    )
    content: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
        comment="章节正文内容",
    )
    chapter_number: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        comment="章节序号 (从 1 开始)",
    )
    status: Mapped[str] = mapped_column(
        String(20),
        default="draft",
        nullable=False,
        comment="章节状态: draft / generating / completed / revised",
    )
    phase4_status: Mapped[str] = mapped_column(
        String(20),
        default="pending",
        nullable=False,
        comment="四阶段精修状态: pending / running / done / failed / skipped",
    )
    word_count: Mapped[int] = mapped_column(
        Integer,
        default=0,
        nullable=False,
        comment="本章字数",
    )
    confirmed_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        comment="确认收纳时间",
    )
    # B3: 生成相关字段
    used_card_ids: Mapped[Optional[list]] = mapped_column(
        JSON,
        nullable=True,
        default=lambda: [],
        comment="使用的卡片 ID 列表",
    )
    generation_mode: Mapped[Optional[str]] = mapped_column(
        String(50),
        nullable=True,
        comment="生成模式: 单卡 / 双卡 / 全选 / 自选+输入",
    )
    generation_prompt: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
        comment="生成提示词",
    )
    generation_weights: Mapped[Optional[dict]] = mapped_column(
        JSON,
        nullable=True,
        comment="卡片权重配置",
    )
    generation_result: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
        comment="生成结果 (LLM 返回的正文)",
    )
    error_message: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
        comment="错误信息",
    )
    retry_count: Mapped[int] = mapped_column(
        Integer,
        default=0,
        nullable=False,
        comment="重试次数",
    )
    generation_duration: Mapped[Optional[int]] = mapped_column(
        Integer,
        nullable=True,
        comment="生成时长 (秒)",
    )

    # ---- Relationships ----
    project = relationship("Project", back_populates="chapters", lazy="selectin")
    dynamic_layer = relationship("DynamicLayer", back_populates="chapter", cascade="all, delete-orphan", lazy="selectin")

    def __repr__(self) -> str:
        return f"<Chapter id={self.id} project={self.project_id} #{self.chapter_number}>"

File: moling-server-rs/crates/moling-db/src/entities/chapter.rs
//! SeaORM entity for Chapter — mirrors Python `app/models/chapter.py`.

use sea_orm::entity::prelude::*;
use sea_orm::prelude::DateTimeUtc;
use serde::{Deserialize, Serialize};
use serde_json::Value as Json;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "chapters")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: String,
    #[sea_orm(column_name = "project_id", indexed)]
    pub project_id: i32,
    pub title: String,
    #[sea_orm(column_type = "Text")]
    pub content: Option<String>,
    pub chapter_number: i32,
    #[sea_orm(default_value = r#""draft""#)]
    pub status: String,
    #[sea_orm(default_value = r#""pending""#)]
    pub phase4_status: String,
    #[sea_orm(default_value = "0")]
    pub word_count: i32,
    pub confirmed_at: Option<DateTimeUtc>,
    pub used_card_ids: Option<Json>,
    pub generation_mode: Option<String>,
    #[sea_orm(column_type = "Text")]
    pub generation_prompt: Option<String>,
    pub generation_weights: Option<Json>,
    #[sea_orm(column_type = "Text")]
    pub generation_result: Option<String>,
    #[sea_orm(column_type = "Text")]
    pub error_message: Option<String>,
    #[sea_orm(default_value = "0")]
    pub retry_count: i32,
    pub generation_duration: Option<i32>,
    #[sea_orm(default_expr = "Expr::current_timestamp()")]
    pub created_at: DateTimeUtc,
    #[sea_orm(default_expr = "Expr::current_timestamp()")]
    pub updated_at: DateTimeUtc,
    #[sea_orm(default_value = "false", indexed)]
    pub is_deleted: bool,
    pub deleted_at: Option<DateTimeUtc>,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::project::Entity",
        from = "Column::ProjectId",
        to = "super::project::Column::Id"
    )]
    Project,
    #[sea_orm(has_many = "super::dynamic_layer::Entity")]
    DynamicLayer,
    #[sea_orm(has_many = "super::generation_task::Entity")]
    GenerationTask,
    #[sea_orm(has_many = "super::draw_history::Entity")]
    DrawHistory,
    #[sea_orm(has_many = "super::vault_changelog::Entity")]
    VaultChangelog,
}

impl ActiveModelBehavior for ActiveModel {}

impl Related<super::project::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Project.def()
    }
}
